### ISSUE TITLE
[DependencyInjection] Overriding services autowired by name under _defaults bind not working

### ISSUE THREAD

**Issue by nico-incubiq** [symfony/symfony#28326]
**Symfony version(s) affected**: 4.1.4

**Description**  
Autowiring allows to bind certain services by name in the *_defaults* section, under *bind* in *services.yaml*. As follow:
```yaml
services:
    _defaults:
        autowire: true
        bind:
            $isDebug: false
```
It is common to override base service definitions in the *services_test.yaml* file for example. So I would have expected to be able to override these default bind too. Like:
```yaml
services:
    _defaults:
        autowire: true
        bind:
            $isDebug: true
```
But if you do so, then the container binding pass will throw the following exception: `Unused binding "$isDebug" in service "App\Tests\Service\MySuperServiceMock".`.

**Possible Solution**  
I tracked it down and this happens because the two definitions of `$isDebug` are considered as distinct bindings. Consequently `ResolveBindingPass::processValue()` will bind the first definition, and the second one remains unbound.  
The solution would be to make the second definition override the first one earlier on in the process, so there is only one binding when getting to `ResolveBindingPass`.

**Workaround**  
My current workaround is to only define the binding in the main service file and make it point to a service alias; so that *services_test.yaml* can override that alias. As follow:
```yaml
services:
    _defaults:
        autowire: true
        bind:
            $superService: '@app.service.super_service'

    app.services.super_service:
        alias: 'App/Service/SuperService'
```

**Comment by GuilhemN** [comment]
Could you share with us more significant parts of your config files please?

Are you trying to override bindings with another config file? Or are you trying to override a service with a different binding?

At least, from what I understand now I think the solution is to be found on your side: we can't make bindings usage detection much more efficient, it is here to help the user detect mistakes and is coded to minimize false positive.

**Comment by stof** [comment]
defaults are per-file. So any default defined in `services_test.yaml` will only affect services defined in `services_test.yaml`.

If what you want is to change the debug flag injected in your normal services, a solution is to use a DI parameter in your binding, and changing the value of the parameter in test (parameters are global in the container):

```yaml
# services.yaml
parameters:
    app.debug: false

services:
    _defaults:
        autowire: true
        bind:
            $isDebug: '%app.debug%'
```


```yaml
# services_test.yaml
parameters:
    app.debug: true
```

**Comment by nico-incubiq** [comment]
Thanks @stof that's indeed the solution I found for the scalar $isDebug. And I also used service aliases as the actual code binds services too. @GuilhemN thanks for your reply too, i'll paste a little more below.
I think I misunderstood the "per-file" bit; for me it meant that services in that file would use the default defined here, but it didn't mean I couldn't override it in a different file, as it is custom to do with everything else in these configuration files. 

So to confirm with you this is the expected behavior; the following situations are currently not valid:
* Explicitly defining an argument already defined as default bind
```yaml
# services.yaml
services:
    _defaults:
        autowire: true
        bind:
            # This is only used by service '@groceries.service', so it's not going to be used anymore in the test environment
            $apiClient: '@eigth_points_guzzle.client.api'
```
```yaml
# services_test.yaml
services:
    groceries.service:
        arguments:
            $apiClient: '@tests.groceries.api_client.mock'
```

* Overriding a default bind when extending the configuration
```yaml
# services.yaml
services:
    _defaults:
        autowire: true
        bind:
            $apiClient: '@eigth_points_guzzle.client.api'
```
```yaml
# services_test.yaml
services:
    _defaults:
        autowire: true
        bind:
            $apiClient: '@tests.groceries.api_client.mock'
```

**Comment by xabbuh** [comment]
Your second example indeed is not supposed to work. In your first example I may misunderstand what you mean is not working here, but if I don't miss anything the `App\Service\Grocery` service should be built as expected as you override the definition coming from the `services.yaml` file here.

**Comment by nico-incubiq** [comment]
That example is working fine on *production* environment, but not in *test* as it complains about the `$apiClient` default bind not being used @xabbuh.

**Comment by oliverde8** [comment]
@nico-incubiq That is not the way default works. Default only works in one file. 

To do what you wish you can simply override the service with another one :
```yml
services: 
    eigth_points_guzzle.client.api: '@test.groceries.api_client.mock'
```
You can create aliases for the service if you need to do it for your own bundle only.

**Comment by nico-incubiq** [comment]
Yeah sure @oliverde8 the initial configurations I tried were not refined or anything. I'd totally be okay with just updating the documentation showing how it needs to be done and saying that all other configuration I proposed is unsupported.

But at the moment I still think think scenario number 1 should work, there's no reason why it would be illegal to override in a different file a service that's originally been defined using a _default.

 * Pushing it a little further, I think example 3 below should be valid too, as the _default bind would definitely be used once in each file and then the service in the test env should override the one in prod env. But this throws the same error `Unused binding "$apiClient" in service ".abstract.instanceof.App\Service\Grocery".`
```php
# services.yaml
services:
    _defaults:
        autowire: true
        bind:
            $apiClient: '@eigth_points_guzzle.client.api'

    App\Service\Grocery:
        # This gets argument $apiClient
```
```php
# services_test.yaml
services:
    _defaults:
        autowire: true
        bind:
            $apiClient: '@test.groceries.api_client.mock'

    App\Service\Grocery:
        # This gets argument $apiClient
```

**Comment by stof** [comment]
@nico-incubiq you are not overriding a default bind. You are defining a different default bind for the other file. But that one will apply only to services defined in that different file.

File-level defaults cannot be overwritten per env (that's why the solution is to make file-level default use some other global features which can be overridden in other files)

**Comment by nicolas-grekas** [comment]
Do we agree it would be better to not throw this exception? I feel like so, isn't it?
I don't know how/if we should fix it, but maybe @nico-incubiq you'd like to create a PR with a test case?

**Comment by stof** [comment]
I think the service override makes the service from the first file disappear from the container before the time where bindings are checked (and so the binding of the first file is not used anymore).

@nicolas-grekas triggering an error for unused file-level bindings might indeed be good for DX when it is actually unused, but I think we have too much false positives. It might indeed be better to remove this error.

**Comment by nico-incubiq** [comment]
Your explanation about service override applying before the bind checks seems very coherent, could that be a bug here?
I would happily try to help with the tests @nicolas-grekas if we agree on the behaviour.

At least the fact it triggers the error made me realise the configuration was not being applied as I expected.

**Comment by GuilhemN** [comment]
> @nicolas-grekas triggering an error for unused file-level bindings might indeed be good for DX when it is actually unused, but I think we have too much false positives. It might indeed be better to remove this error.

Imo it does more good than harm but that's subjective :)
Whatever you decide I don't think we should make this error more accurate, the only way I see is remembering all removed services containing bindings and checking if they're not using them, but that's very risky, the definition may have changed or be broken so, most likely not something we want to do.

**Comment by nico-incubiq** [comment]
Your explanation about service override applying before the bind checks seems very coherent, could that be a bug here?
I would happily try to help with the tests @nicolas-grekas if we agree on the behaviour.

At least the fact it triggers the error made me realise the configuration was not being applied as I expected.

**Comment by nicolas-grekas** [comment]
I agree this exception is nice, I was just wondering if we could change the behavior in this very specific context, *if* possible yes. @stof I'm not sure what you mean with "we have too much false positives". If you're thinking about more, we should find a way to be explicit. Like @GuilhemN, I tend to think catching errors (typos, etc.) is a good thing here by default.

**Comment by GuilhemN** [comment]
@nicolas-grekas we could also mark bindings of services removed as used (without actually checking) but that could potentially create a lot of false positives.

**Comment by GuilhemN** [comment]
In the `ContainerBuilder::removeDefinition` method and when overriding a service we could keep a list of bindings those services used (bindings have a unique id, shared for defaults), that would then be passed to the bindings compiler pass in order to ignore them if they're not used elsewhere.

**Comment by stof** [comment]
@GuilhemN but until autowiring is performed, a service does not use any binding. It has a list of bindings it **can** use. If the only fact of being able to use a binding (even if you don't actually need it) marks it as used, this error would be useless, as it would only happen if you define a default binding in a file which does not declare any service at all.

**Comment by GuilhemN** [comment]
@stof only bindings from removed services would not longer be checked (and I don't think services are often removed before bindings are processed, so it shouldn't happen that often), most bindings would be checked as they are now but indeed my proposal isn't perfect and if you remove one service in a file, all its defaults bindings will no longer be checked but I believe it's better than just removing the error while still fixing this issue.

**Comment by nicolas-grekas** [comment]
Looks like we an idea worth trying IMHO!

**Comment by GuilhemN** [comment]
I won't have time to do it myself sorry, but I'd be pleased to answer any question about my proposal if someone wants to give it a try :)

**Comment by gonzalovilaseca** [comment]
@GuilhemN  I can give it a try, it would be much easier if you can provide failing tests :-), if you can't, no problem, I'll try to do them.

**Comment by GuilhemN** [comment]
If you take these two files, an unexpected exception will be thrown:
```yml
# /src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/defaults_bindings.yml
services:
    _defaults:
        bind: { $quz: value }
    
    bar:
        class: Symfony\Component\DependencyInjection\Tests\Fixtures\Bar

    foo:
        class: Symfony\Component\DependencyInjection\Tests\Fixtures\FactoryDummy
```
```yml
# /src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/defaults_bindings2.yml
services:
    _defaults:
        bind: { $quz: ~ }

    bar:
        class: Symfony\Component\DependencyInjection\Tests\Fixtures\Bar
```
```php
# /src/Symfony/Component/DependencyInjection/Tests/Loader/YamlFileLoaderTest.php
class .. {
    public function testOverriddenDefaultsBindings()
    {
        $container = new ContainerBuilder();
        $loader = new YamlFileLoader($container, new FileLocator(self::$fixturesPath.'/yaml'));
        $loader->load('defaults_bindings.yml');
        $loader->load('defaults_bindings2.yml');
    }

**Comment by nicolas-grekas** [comment]
implemented in #29597 if you want to have a look.

**Comment by przemyslaw-bogusz** [comment]
I will gladly give it a shot. I will make a PR once the reverting PR is merged (https://github.com/symfony/symfony/pull/29853). I see it failed the coding standards check, due to the new short array syntax rule.
